README.md
[jsRequest](https://github.com/Panamark/jsRequest)
===================================================

A JavaScript framework for fast and easy asynchronous JavaScript files loading

Copyright (C) 2013 [Mark Linus](https://github.com/MarkLinus/) and [The Panuto](https://github.com/ThePanuto/)

### Quick Index

 - [About](#about)
   - [The loading process](#the-loading-process)
   - [Callbacks](#callbacks)
 - [Usage & Syntax](#usage--syntax)
   - [The `options` argument](#the-options-argument)
   - [Interactive loading - The `progress` event](#interactive-loading---the-progress-event)
   - [Being pacient - The `.wait` method](#being-pacient---the-wait-method)
 - [Other Features](#other-features)
   - [History](#history)
   - [Files information](#files-information)
 - [Compatibility](#compatibility)
 - [Known bugs](#known-bugs)
 - [Authors](#authors)
 - [License](#license)

- - -

About
=====

jsRequest is a crossbrowser JavaScript framework that allows programmers to include external JavaScript
files in their pages with nothing more than a single line.
Its flexibility and high-compatibility ensures that your files will be loaded even on older browsers.

### The loading process

It uses two methods to load files:

 1. Ajax (XMLHttpRequest): The most rich and reliable method, makes an Ajax `GET` call to the file's
	url and, once it's totally received, appends a new `<script>` tag to the `<head>` and sets the response
	as its content.

 2. `<script>` tag: An alternative for older browsers, files stored outside the current server and files loaded
	via the `file://` protocol, it consists in the tradicional method of inserting a new `<script>` tag to the
	`<head>`, which `src` attribute's is set to the file's url.

### Callbacks

This framework allows the programmer to define specific callbacks to each request. These callback are:

 * Success: Called if the file is loaded successfully
	
 * Failure: Called if the an error occurs during the loading proccess
	
 * Progress: Called whenever a portion of the file is loaded. Thus, it's called at least once during the
   loading file

Usage & Syntax
==============

It's very simple to use jsRequest. First, you need to include it in your page:
```html
<script type="text/javascript" src="jsRequest-1.0.1.min"></script>
```

Then you just need to invoke its methods. In most cases you won't need more than ten lines of code, like
the example below:
```javascript
jsRequest.load("myScript.js");
```

Now, let's suppose you want to load an external script like jQuery, or even a script under the `file://`
protocol. No headaches! Just type it's url:
```javascript
jsRequest.load("http://code.jquery.com/jquery-1.10.2.js");
jsRequest.load("file:///C:/myFile.js");
```
Done! jsRequest will do all the hard work for you.
However, you may also want to know whenever the download fails or not. That's why jsRequest allows you to
configure some options. Just keep reading and see below "The `options` argument" section.

> Tip: Since each method of jsRequest returns the `jsRequest` object itself, it's possible to nestle all methods
> in a single statement, like below:
> ```javascript
> jsRequest
> 	.load("myScript.js")
> 	.load("http://code.jquery.com/jquery-1.10.2.js")
> 	.load("file:///C:/myFile.js");
> ```
> The example above is the pretty exact as the following:
> ```javascript
> jsRequest.load("myScript.js");
> jsRequest.load("http://code.jquery.com/jquery-1.10.2.js");
> jsRequest.load("file:///C:/myFile.js");
> ```
>
> You may also make a single call of the `.load` method:
> ```javascript
> jsRequest.load(
> 		"myScript.js",
> 		"http://code.jquery.com/jquery-1.10.2.js",
> 		"file:///C:/myFile.js"
> 	);
> ```

### The `options` argument

As mentioned in "Callbacks" and "Usage & Syntax" sections, the method `.load` allows you to define callbacks
for each requested file. These callbacks can e compared to the DOM events `onload` (success) and `onerror`
(failure). You can do that in a pretty simple syntax, like JSON:
```javascript
{
	success: function (url, xhr, method, event) {
		// What happens if the file is loaded
	},
	failure: function (url, xhr, method, event) {
		// What happens if the loading failed
	}
}
```

And then just pass it as the last argument of the `.load` method.
Do you remember the example of jQuery? Well, let's how it would be with some options:
```javascript
jsRequest.load("http://code.jquery.com/jquery-1.10.2.js", {
	success: function () {
		alert("Hello, John!");
	},
	failure: function () {
		alert("Sorry, I can't live without jQuery :c");
	}
});
```

If you just want provide the `success` callback, you can simply pass a function instead of the `options`
object:
```javascript
jsRequest.load("myFile.js", function () {
	alert("My file is loaded");
});
```

> Tip: All properties of `options` are optional.

- - -

> Tip: Since you can make a single call for more than one file, you may also do the following:
> ```javascript
> jsRequest.load(
> 	"myScript.js",
> 	"http://code.jquery.com/jquery-1.10.2.js",
> 	"file:///C:/myFile.js",
> 	{
> 		success: function (url) {
> 			alert("Yay! " + url + " is loaded");
> 		},
> 		failure: function (url) {
> 			alert("Couldn't load " + url + " =/");
> 		}
> 	}
> );
> ```
> This way, the `options` argument will be given to each request.

### Interactive loading - The `progress` event

So far, we've learned about `success` and `failure` callbacks. Now we will see the third method of the
`options` argument: the `progress`. This one is called whenever the download advances, which allows the user
to use its creativity to create a custom loading bar, for instance. You just need to add this method to the
`options` argument like described below:
```javascript
{
	success: ...,
	failure: ...,
	progress: function (url, percentage, xhr, method, event) {
		console.log(percentage + "% loaded");
	}
}
```

### Being pacient - The `.wait` method

Another feature of jsRequest is the method `.wait`, which is similar to the `success` callback. This method
takes a single argument, which is a function that will be called right after the last file requested is loaded.
Thanks to this, we can have a request like this:
```javascript
jsRequest
	.load("http://code.jquery.com/jquery-1.10.2.js")
	.wait(function () {
		jsRequest.load("myjQueryPlugin.js", function () {
			alert("Everything is loaded ^^");
		});
	});
```

The good thing of this method is that, if the first argument is a string, then  jsRequest assumes that the programmer
is making a new request, so it behaves like `.load`, which means that the all arguments are urls and the last is
the `options` (if it's not a string). Therefore:
```javascript
jsRequest
	.load("http://code.jquery.com/jquery-1.10.2.js")
	.wait("myjQueryPlugin.js", function () {
		alert("Everything is loaded ^^");
	});
```
Is also valid.

> Tip: If `.wait` is called before the page loads and no request was made before, then it works like the `document`'s
> `onload` event: 
> ```javascript
> jsRequest
> 	.wait(function () {
> 		alert("Page loaded :)");
> 	})
>	.wait("myScript.js");	// loads 'myScript.js' once the page is loaded
> ```

Other Features
==============

### History
jsRequest also stores a request history, where all requested files are stored in array, which is ordered by request
date. Each element of this array is an object that stores the file's url and size and the request's starting date,
ending date and status. This object is described below:
```javascript
{
	/* number */ endDate: ...
	/* number */ size: ...,
	/* number */ startDate: ...,
	/* string */ state: ...,
	/* string */ url: ...
}
```
The history is a property of `jsRequest` can be accessed by `jsRequest.history`.

### Files information

Very similar to the history, the files information is an object where each requested and successfully downloaded file
is an associative index (aka "key") of this object, and each element is an object with the same structure described
in the "History" section.
The files information can be accessed by `jsRequest.files` and alllows the programmer to know some basic information
about each requested file, like this example:
```javascript
jsRequest.load("myScript.js", function (url) {
	alert(url + " took " + (jsRequest.files[url].endDate - jsRequest.files[url].startDate) + " ms to be loaded");
});
```

Compatibility
=============

jsRequest was tested and worked on these browsers:

 * Google Chrome 28.0
 * Mozilla Firefox 22.0
 * Opera 15.0
 * Safari 5.1
 * Internet Explorer 5+

It may also work on other browsers or older versions of the mentioned above. However, they weren't tested yet.

Known bugs
==========

Since IE \<9 doesn't support the `onload` and `onerror` events for `<script>` tags, the `failure` callback
doesn't work on these versions, and therefore it always calls the `success` callback.
Thus, it's recomended to create a custom validator function, like this:
```javascript
jsRequest
	.load("http://code.jquery.com/jquery-1.10.2.js", function () {
		if (window.jQuery) {
			jsRequest.load("myjQueryPlugin.js", function () {
				alert("Everything is loaded ^^");
			});
		} else {
			alert("Sorry, I couldn't find jQuery");
		}
	});
```

Authors
=======
This project was developed and is currently being maintened by [Mark Linus](https://github.com/MarkLinus/)
and [The Panuto](https://github.com/ThePanuto/)

License
=======

See [LICENSE.md](https://github.com/Panamark/jsRequest/blob/master/LICENSE.md)


CHANGELOG.md
- 1.0.1
 - Fixed [issue #1](https://github.com/Panamark/jsRequest/issues/1)

- 1.0.0
  - First release


src/jsRequest-1.0.0.js
(function (window, waitForBody) {
	'use strict';

	var doc = window.document;
	var isModule = typeof (module) === 'object' && !!module && typeof (module.exports) === 'object';

	/*
	 * XMLHttpRequest compatibility
	 */
	var newXhr = (function () {
		if (!!window.XMLHttpRequest) {
			return function () {
				return new window.XMLHttpRequest();
			};
		}
		if (!!window.activeXObject) {
			return function () {
				return new window.ActiveXObject('Microsoft.XMLHTTP');
			};
		}
		return function () {
			return null;
		};
	})();

	/*
	 * Private
	 */
	var tasks = [];
	var callbacks = {};
	var pageCallbacks = [];	// Called when the page is loaded (see function change in the bottom)
	var lastAdded = null;
	var history = [];
	var files = {};

	var isFn = function (fn) {
		return fn instanceof Function;
	};

	var attachEvent = function (target, event, action) {
		event = 'on' + event;
		target[event] = !isFn(target[event]) ? action : function () {
			try {
				target[event].apply(this, arguments);
			} catch (ex) {
				console.error(ex.stack);
			}
			action.apply(this, arguments);
		};
	};

	var addEvent = (function () {
		if (!!window.addEventListener) {
			return function (target, event, action) {
				if (!target) {
					return null;
				}
				return target.addEventListener ? target.addEventListener(event, action, false) : attachEvent(target, event, action);
			};
		}
		if (window.attachEvent) {
			return function (target, event, action) {
				if (!target) {
					return null;
				}
				return target.attachEvent ? target.attachEvent('on' + event, action, false) : attachEvent(target, event, action);
			};
		}
		return attachEvent;
	})();

	var isReady = (function () {
		if(isModule) {
			return function () {
				return true;
			};
		}
		if(!!waitForBody) {
			return function () {
				return doc.readyState === 'complete' && doc.body !== null;
			};
		}
		return function () {
			return doc.readyState === 'complete';
		};
	})();

	var scheduleTask = function (task) {
		if (isReady()) {
			task.call();
		} else {
			tasks.push(task);
		}
	};

	var runCallbacks = function (callbacks, context, args) {
		while (callbacks.length > 0) {
			try {
				callbacks.shift().apply(context, args);
			} catch (ex) {
				console.error(ex.stack);
			}
		}
	};

	var runScript = function (code) {
		var script = doc.createElement('script');
		script.setAttribute('type', 'text/javascript');
		script.text = code;
		doc.getElementsByTagName('head')[0].appendChild(script);
	};

	var parseURL = function (url) {
		if (typeof (url) !== 'string') {
			url = '';
		}
		var a = doc.createElement('a');
		a.href = url;
		var pieces = {
			hash	: a.hash,
			host	: a.host,
			hostname: a.hostname,
			origin	: a.origin,
			pathname: a.pathname,
			port	: a.port,
			protocol: a.protocol,
			query	: a.query
		};
		var l = window.location;
		return {
			hash	: !pieces.hash		? l.hash		: pieces.hash,
			host	: !pieces.host		? l.host		: pieces.host,
			hostname: !pieces.hostname	? l.hostname	: pieces.hostname,
			origin	: !pieces.origin	? l.origin		: pieces.origin,
			pathname: !pieces.pathname	? l.pathname	: pieces.pathname,
			port	: !pieces.port		? l.port		: pieces.port,
			protocol: !pieces.protocol	? l.protocol	: pieces.protocol,
			query	: !pieces.query		? l.query		: pieces.query
		};
	};

	var jsRequest;

	/*
	 * Loads the script from a script tag
	 */
	var scriptTagTask = function (url, options) {
		return function () {
			var startDate = new Date().getTime();
			var script = doc.createElement('script');
			script.setAttribute('type', 'text/javascript');
			script.setAttribute('src', url);
			script.setAttribute('async', true);
			var info = {
				url: url,
				startDate: startDate,
				state: 'failure',
				size: 0
			};
			/*
			 * User-defined events
			 */
			var success = function (e) {
				/*
				 * Progress
				 */
				try {
					options.progress.apply(jsRequest, [url, 100, null, 'ajax', e]);
				} catch (ex) {
					console.error(ex.stack);
				}
				/*
				 * Success
				 */
				info.state = 'success';
				info.endDate = new Date().getTime();
				history.push(info);
				files[url] = info;
				var args = [url, null, 'scripttag', e];
				runCallbacks(callbacks[url], jsRequest);
				try {
					options.success.apply(jsRequest, args);
				} catch (ex) {
					console.error(ex.stack);
				}
			};
			var failure = function (e) {
				/*
				 * Failure
				 */
				info.state = 'failure';
				info.endDate = new Date().getTime();
				history.push(info);
				files[url] = info;
				var args = [url, null, 'scripttag', e];
				try {
					options.failure.apply(jsRequest, args);
				} catch (ex) {
					console.error(ex.stack);
				}
			};
			var hasOnloadEvent = navigator.appName !== 'Microsoft Internet Explorer' || !!window.addEventListener;
			if (hasOnloadEvent) {
				/*
				 * For good browsers only
				 */
				addEvent(script, 'load', success);
				addEvent(script, 'error', failure);
			} else {
				/*
				 * For bad browsers like IE <9 - Doesn't support 'onerror' event
				 */
				var fired = false;
				addEvent(script, 'readystatechange', function (e) {
					if (!fired && (script.readyState === 'loaded' || script.readyState === 'complete')) {
						fired = true;
						success.call(script, e);
					}
				});
			}
			doc.getElementsByTagName('head')[0].appendChild(script);
		};
	};

	/*
	 * Loads the JavaScript file from an AJAX call
	 */
	var ajaxTask = function (url, options) {
		var xhr = newXhr();
		if (xhr === null) {
			return scriptTagTask(url, options);
		}
		return function () {
			var startDate = new Date().getTime();
			xhr.open('GET', url);
			xhr.setRequestHeader('Content-Type', 'text/javascript; charset=utf-8');
			if (!!xhr.overrideMimeType) {
				xhr.overrideMimeType('text/javascript; charset=utf-8');
			}
			/*
			 * User-defined events
			 */
			var total = 0;
			addEvent(xhr, 'readystatechange', function (e) {
				if (xhr.readyState === 2) {
					total = parseInt(xhr.getResponseHeader('Content-Length') || '0');
				} else if (xhr.readyState === 3) {
					/*
					 * Progress
					 */
					try {
						options.progress.apply(jsRequest, [url, 100 * xhr.response.length / total, xhr, 'ajax', e]);
					} catch (ex) {
						console.error(ex.stack);
					}
				} else if (this.readyState === 4) {
					var success = (this.status >= 200 && this.status < 300) || this.status === 304;
					var info = {
						url: url,
						startDate: startDate,
						endDate: new Date().getTime(),
						state: success ? 'success' : 'failure',
						size: total
					};
					history.push(info);
					files[url] = info;
					var args = [url, xhr, 'ajax', e];
					if (success) {
						/*
						 * Success
						 */
						runScript(this.response || this.responseText);
						runCallbacks(callbacks[url], jsRequest);
						try {
							options.success.apply(jsRequest, args);
						} catch (ex) {
							console.error(ex.stack);
						}
					} else {
						/*
						 * Failure
						 */
						try {
							options.failure.apply(jsRequest, args);
						} catch (ex) {
							console.error(ex.stack);
						}
					}
				}
			});
			xhr.send();
		};
	};

	/*
	 * Validates options. If the second argument is an function, it's interpreted as
	 * the 'success' function. Thus when we receive something like this:
	 * 	jsRequest.load('path/to/file.js',function(){
	 * 		console.log('success');
	 * 	});
	 * We assume the user meant something like this:
	 * 	jsRequest.load('path/to/file.js',{
	 * 		success: function(){
	 * 			console.log('success');
	 * 		}
	 * 	});
	*/
	var normOptions = function (options) {
		if (isFn(options)) {
			return {
				success: options,
				failure: function(){},
				progress: function(){}
			}
		}
		options = options || {};
		return {
			success: isFn(options.success) ? options.success : function () {},
			failure: isFn(options.failure) ? options.failure : function () {},
			progress: isFn(options.progress) ? options.progress : function () {}
		};
	};

	/*
	 * The jsRequest object
	 */
	jsRequest = {
		/*
		 * Main information about jsRequest and its authors
		 */
		about: {
			NAME: 'jsRequest',
			VERSION: '1.0.0',
			DATE: '2013-07-14 (yyyy-mm-dd)',
			DESCRIPTION: 'A JavaScript framework for fast and easy asynchronous JavaScript files loading',
			AUTHORS: [
				{
					NAME: 'Danilo Marcolino Valente',
					NICK: 'Mark Linus',
					EMAIL: 'dani_lovalente@hotmail.com',
					PAGE: 'https://github.com/MarkLinus'
				},
				{
					NAME: 'Bruno Lu\u00EDs Panuto Silva',
					NICK: 'The Panuto',
					EMAIL: 'bl.panuto@gmail.com',
					PAGE: 'https://github.com/ThePanuto'
				}
			],
			COPYRIGHT: '(C)2013 by Danilo Marcolino Valente and Bruno Lu\u00EDs Panuto Silva'
		},
		/*
		 * An array containing information about each loaded file, sorted by time
		 */
		history: history,
		/*
		 * An associative array containing information about each loaded file
		 */
		files: files,
		/*
		 * Load's a JavaScript file in two different ways:
		 * - Inserts a script tag that receives the file's url as value of the "src" parameter
		 * if the requested file is in another server or the current page is running under
		 * the "file://" protocol
		 * - Makes an Ajax call and executes the returned content as JavaScript script if the
		 * requested file is stored in the same server of the current page
		 *
		 * The first argument is always a string containing the url of the requested file. However,
		 * the second one can be both an object that contains a set of options or a function called
		 * as callback of the "success" state.
		 *
		 * The options argument is described below:
		 * {
		 *     success: function(string url, XMLHttpRequest xhr, string method, XMLHttpRequestProgressEvent event),
		 *     failure: function(string url, XMLHttpRequest xhr, string method, XMLHttpRequestProgressEvent event),
		 *     progress: function(string url, XMLHttpRequest xhr, string method, XMLHttpRequestProgressEvent event)
		 * }
		 */
		load: function () {
			var argc = arguments.length;
			var options = normOptions(argc>1 && typeof (argc - 1) !== 'string' ? arguments[--argc] : null);
			for (var i = 0; i < argc; i++) {
				var url = arguments[i];
				lastAdded = url;
				if (!callbacks[url]) {
					callbacks[url] = [];
				}
				if (typeof (url) === 'string') {
					scheduleTask(window.location.protocol === 'file:' || window.location.hostname !== parseURL(url).hostname
						? scriptTagTask(url, options)
						: ajaxTask(url, options)
					);
				}
			}
			return jsRequest;
		},
		/*
		 * Executes a callback function right after the last requested file is loaded.
		 * If the provided action is actually a string, then we assume it's an url and
		 * the user wants to load it. So, 'jsRequest.wait("foo/bar.js");' is
		 * exactly the same of jsRequest.wait(function(){ jsRequest.load("foo/bar.js"); });.
		 * Also, in this case we consider that the user may have also provided a second
		 * argument, which is passed to jsRequest.load as the options argument.
		 * There is also a third function overload, which takes two arguments, the file's
		 * url and a function, which is considered as the callback in case of success.
		 * Thus, jsRequest.load("foo/bar.js", function(){}); is the same of
		 * jsRequest.load("foo/bar.js", {success: function(){}});.
		 */
		wait: function (action) {
			if (typeof (action) === 'string') {
				var args = arguments;
				action = function () {
					jsRequest.load.apply(this, args);
				}
			}
			if (!isFn(action)) {
				action = function () {};
			}
			var target = lastAdded;
			if (target === null) {
				if (isReady()) {
					runCallbacks([action], jsRequest);
				} else {
					pageCallbacks.push(action);
				}
			} else {
				callbacks[target].push(action);
			}
			return jsRequest;
		}
	};

	/*
	 * Called once the page is loaded
	 */
	var init = function () {
		runCallbacks(pageCallbacks, jsRequest);
		while (tasks.length > 0) {
			tasks.shift().call();
		}
	};

	/*
	 * Expose
	 */
	if (isModule) {
		module.exports = jsRequest;
		init();
	} else {
		window.jsRequest = jsRequest;
		/*
		 * Wait for page loading
		 */
		var isPageLoaded = false;
		var change = function () {
			if (isReady() && !isPageLoaded) {
				isPageLoaded = true;
				init();
			}
		};
		addEvent(window, 'load', change);
		addEvent(doc, 'load', change);
		addEvent(doc, 'readystatechange', change);
		change();	// If the page is already loaded, the events above won't be fired
	}
})(window, true);


src/jsRequest-1.0.1.js
(function (window, waitForBody) {
	'use strict';
	
	var doc = window.document;
	var isModule = typeof (module) === 'object' && !!module && typeof (module.exports) === 'object';
	
	/*
	 * XMLHttpRequest compatibility
	 */
	var newXhr = (function () {
		if (!!window.XMLHttpRequest) {
			return function () {
				return new window.XMLHttpRequest();
			};
		}
		if (!!window.activeXObject) {
			return function () {
				return new window.ActiveXObject('Microsoft.XMLHTTP');
			};
		}
		return function () {
			return null;
		};
	})();
	
	/*
	 * jsRequest
	 */
	var tasks = [];
	var callbacks = {};
	var pageCallbacks = [];	// Called when the page is loaded (see function change in the bottom)
	var lastAdded = null;
	var history = [];
	var files = {};
	
	var isFn = function (fn) {
		return fn instanceof Function;
	};
	
	var attachEvent = function (target, event, action) {
		event = 'on' + event;
		target[event] = !isFn(target[event]) ? action : function () {
			try {
				target[event].apply(this, arguments);
			} catch (ex) {
				console.error(ex.stack);
			}
			action.apply(this, arguments);
		};
	};
	
	var addEvent = (function () {
		if (!!window.addEventListener) {
			return function (target, event, action) {
				if (!target) {
					return null;
				}
				return target.addEventListener ? target.addEventListener(event, action, false) : attachEvent(target, event, action);
			};
		}
		if (window.attachEvent) {
			return function (target, event, action) {
				if (!target) {
					return null;
				}
				return target.attachEvent ? target.attachEvent('on' + event, action, false) : attachEvent(target, event, action);
			};
		}
		return attachEvent;
	})();
	
	var isReady = (function () {
		if(isModule) {
			return function () {
				return true;
			};
		}
		if(!!waitForBody) {
			return function () {
				return doc.readyState === 'complete' && doc.body !== null;
			};
		}
		return function () {
			return doc.readyState === 'complete';
		};
	})();
	
	var scheduleTask = function (task) {
		if (isReady()) {
			task.call();
		} else {
			tasks.push(task);
		}
	};
	
	var runCallbacks = function (callbacks, scope, args) {
		while (callbacks.length > 0) {
			try {
				callbacks.shift().apply(scope, args);
			} catch (ex) {
				console.error(ex.stack);
			}
		}
	};
	
	var runScript = function (code) {
		var script = doc.createElement('script');
		script.setAttribute('type', 'text/javascript');
		script.text = code;
		doc.getElementsByTagName('head')[0].appendChild(script);
	};
	
	var parseURL = function (url) {
		if (typeof (url) !== 'string') {
			url = '';
		}
		var a = doc.createElement('a');
		a.href = url;
		var pieces = {
			hash	: a.hash,
			host	: a.host,
			hostname: a.hostname,
			origin	: a.origin,
			pathname: a.pathname,
			port	: a.port,
			protocol: a.protocol,
			query	: a.query
		};
		var l = window.location;
		return {
			hash	: !pieces.hash		? l.hash		: pieces.hash,
			host	: !pieces.host		? l.host		: pieces.host,
			hostname: !pieces.hostname	? l.hostname	: pieces.hostname,
			origin	: !pieces.origin	? l.origin		: pieces.origin,
			pathname: !pieces.pathname	? l.pathname	: pieces.pathname,
			port	: !pieces.port		? l.port		: pieces.port,
			protocol: !pieces.protocol	? l.protocol	: pieces.protocol,
			query	: !pieces.query		? l.query		: pieces.query
		};
	};
	
	var jsRequest;
	
	/*
	 * Loads the script from a script tag
	 */
	var scriptTagTask = function (url, options) {
		return function () {
			var startDate = new Date().getTime();
			var script = doc.createElement('script');
			script.setAttribute('type', 'text/javascript');
			script.setAttribute('src', url);
			script.setAttribute('async', true);
			var info = {
				url: url,
				startDate: startDate,
				state: 'failure',
				size: 0
			};
			/*
			 * User-defined events
			 */
			var success = function (e) {
				/*
				 * Progress
				 */
				try {
					options.progress.apply(jsRequest, [url, 100, null, 'ajax', e]);
				} catch (ex) {
					console.error(ex.stack);
				}
				/*
				 * Success
				 */
				info.state = 'success';
				info.endDate = new Date().getTime();
				history.push(info);
				files[url] = info;
				var args = [url, null, 'scripttag', e];
				runCallbacks(callbacks[url], jsRequest);
				try {
					options.success.apply(jsRequest, args);
				} catch (ex) {
					console.error(ex.stack);
				}
			};
			var failure = function (e) {
				/*
				 * Failure
				 */
				info.state = 'failure';
				info.endDate = new Date().getTime();
				history.push(info);
				files[url] = info;
				var args = [url, null, 'scripttag', e];
				try {
					options.failure.apply(jsRequest, args);
				} catch (ex) {
					console.error(ex.stack);
				}
			};
			var hasOnloadEvent = navigator.appName !== 'Microsoft Internet Explorer' || !!window.addEventListener;
			if (hasOnloadEvent) {
				/*
				 * For good browsers only
				 */
				addEvent(script, 'load', success);
				addEvent(script, 'error', failure);
			} else {
				/*
				 * For bad browsers like IE <9 - Doesn't support 'onerror' event
				 */
				var fired = false;
				addEvent(script, 'readystatechange', function (e) {
					if (!fired && (script.readyState === 'loaded' || script.readyState === 'complete')) {
						fired = true;
						success.call(script, e);
					}
				});
			}
			doc.getElementsByTagName('head')[0].appendChild(script);
		};
	};
	
	/*
	 * Loads the JavaScript file from an AJAX call
	 */
	var ajaxTask = function (url, options) {
		var xhr = newXhr();
		if (xhr === null) {
			return scriptTagTask(url, options);
		}
		return function () {
			var startDate = new Date().getTime();
			xhr.open('GET', url);
			xhr.setRequestHeader('Content-Type', 'text/javascript; charset=utf-8');
			if (!!xhr.overrideMimeType) {
				xhr.overrideMimeType('text/javascript; charset=utf-8');
			}
			/*
			 * User-defined events
			 */
			var total = 0;
			addEvent(xhr, 'readystatechange', function (e) {
				if (xhr.readyState === 2) {
					total = parseInt(xhr.getResponseHeader('Content-Length') || '0');
				} else if (xhr.readyState === 3) {
					/*
					 * Progress
					 */
					try {
						options.progress.apply(jsRequest, [url, 100 * xhr.response.length / total, xhr, 'ajax', e]);
					} catch (ex) {
						console.error(ex.stack);
					}
				} else if (this.readyState === 4) {
					var success = (this.status >= 200 && this.status < 300) || this.status === 304;
					var info = {
						url: url,
						startDate: startDate,
						endDate: new Date().getTime(),
						state: success ? 'success' : 'failure',
						size: total
					};
					history.push(info);
					files[url] = info;
					var args = [url, xhr, 'ajax', e];
					if (success) {
						/*
						 * Success
						 */
						runScript(this.response || this.responseText);
						runCallbacks(callbacks[url], jsRequest);
						try {
							options.success.apply(jsRequest, args);
						} catch (ex) {
							console.error(ex.stack);
						}
					} else {
						/*
						 * Failure
						 */
						try {
							options.failure.apply(jsRequest, args);
						} catch (ex) {
							console.error(ex.stack);
						}
					}
				}
			});
			xhr.send();
		};
	};
	
	/*
	 * Validates options. If the second argument is an function, it's interpreted as
	 * the 'success' function. Thus when we receive something like this:
	 * 	jsRequest.load('path/to/file.js',function(){
	 * 		console.log('success');
	 * 	});
	 * We assume the user meant something like this:
	 * 	jsRequest.load('path/to/file.js',{
	 * 		success: function(){
	 * 			console.log('success');
	 * 		}
	 * 	});
	*/
	var normOptions = function (options) {
		if (isFn(options)) {
			return {
				success: options,
				failure: function(){},
				progress: function(){}
			}
		}
		options = options || {};
		return {
			success: isFn(options.success) ? options.success : function () {},
			failure: isFn(options.failure) ? options.failure : function () {},
			progress: isFn(options.progress) ? options.progress : function () {}
		};
	};
	
	/*
	 * The jsRequest object
	 */
	jsRequest = {
		/*
		 * Main information about jsRequest and its authors
		 */
		about: {
			NAME: 'jsRequest',
			VERSION: '1.0.1',
			DATE: '2013-07-20 (yyyy-mm-dd)',	// yyyy-mm-dd
			DESCRIPTION: 'A JavaScript framework for fast and easy asynchronous JavaScript files loading',
			AUTHORS: [
				{
					NAME: 'Danilo Marcolino Valente',
					NICK: 'Mark Linus',
					EMAIL: 'dani_lovalente@hotmail.com',
					PAGE: 'https://github.com/MarkLinus'
				},
				{
					NAME: 'Bruno Lu\u00EDs Panuto Silva',
					NICK: 'The Panuto',
					EMAIL: 'bl.panuto@gmail.com',
					PAGE: 'https://github.com/ThePanuto'
				}
			],
			COPYRIGHT: '(C)2013 by Danilo Marcolino Valente and Bruno Lu\u00EDs Panuto Silva'
		},
		/*
		 * An array containing information about each loaded file, sorted by time
		 */
		history: history,
		/*
		 * An associative array containing information about each loaded file
		 */
		files: files,
		/*
		 * Load's a JavaScript file in two different ways:
		 * <ul>
		 * <li>Inserts a script tag that receives the file's url as value of the "src" parameter
		 * if the requested file is in another server or the current page is running under
		 * the "file://" protocol</li>
		 * <li>Makes an Ajax call and executes the returned content as JavaScript script if the
		 * requested file is stored in the same server of the current page</li>.
		 *
		 * The first argument is always a string containing the url of the requested file. However,
		 * the second one can be both an object that contains a set of options or a function called
		 * as callback of the "success" state.
		 *
		 * The options argument is described below:
		 * {
		 *     success: function(string url, XMLHttpRequest xhr, string method, XMLHttpRequestProgressEvent event),
		 *     failure: function(string url, XMLHttpRequest xhr, string method, XMLHttpRequestProgressEvent event),
		 *     progress: function(string url, XMLHttpRequest xhr, string method, XMLHttpRequestProgressEvent event)
		 * }
		 */
		load: function () {
			var argc = arguments.length;
			var options = normOptions(argc>1 && typeof (argc - 1) !== 'string' ? arguments[--argc] : null);
			for (var i = 0; i < argc; i++) {
				var url = arguments[i];
				lastAdded = url;
				if (!callbacks[url]) {
					callbacks[url] = [];
				}
				if (typeof (url) === 'string') {
					scheduleTask(window.location.protocol === 'file:' || window.location.hostname !== parseURL(url).hostname
						? scriptTagTask(url, options)
						: ajaxTask(url, options)
					);
				}
			}
			return jsRequest;
		},
		/*
		 * Executes a callback function right after the last requested file is loaded.
		 * If the provided action is actually a string, then we assume it's an url and
		 * the user wants to load it. So, <pre>jsRequest.wait("foo/bar.js");</pre> is
		 * exactly the same of <pre>jsRequest.wait(function(){ jsRequest.load("foo/bar.js"); });</pre>.
		 * Also, in this case we consider that the user may have also provided a second
		 * argument, which is passed to jsRequest.load as the options argument.
		 * There is also a third function overload, which takes two arguments, the file's
		 * url and a function, which is considered as the callback in case of success.
		 * Thus, <pre>jsRequest.load("foo/bar.js", function(){});</pre> is the same of
		 * jsRequest.load("foo/bar.js", {success: function(){}});</pre>.
		 */
		wait: function (action) {
			/*
			 * Issue #1:
			 * - Bug: Changing the original reference of 'action' also changes the original
			 *     reference of 'arguments[0]' in the non-strict mode. Thus, 'arguments[0]'
			 *     would be a function rather than a string
			 * - Solution: Created a var called '_action' instead of changing the original
			 *     reference of 'action'
			 */
			var _action = action;
			if (typeof (_action) === 'string') {
				var args = arguments;
				_action = function () {
					jsRequest.load.apply(this, args);
				}
			}
			if (!isFn(_action)) {
				_action = function () {};
			}
			var target = lastAdded;
			if (target === null) {
				if (isReady()) {
					runCallbacks([_action], jsRequest);
				} else {
					pageCallbacks.push(_action);
				}
			} else {
				callbacks[target].push(_action);
			}
			return jsRequest;
		}
	};
	
	/*
	 * Called once the page is loaded
	 */
	var init = function () {
		runCallbacks(pageCallbacks, jsRequest);
		while (tasks.length > 0) {
			tasks.shift().call();
		}
	};
	
	/*
	 * Expose
	 */
	if (isModule) {
		module.exports = jsRequest;
		init();
	} else {
		window.jsRequest = jsRequest;
		/*
		 * Wait for page loading
		 */
		var isPageLoaded = false;
		var change = function () {
			if (isReady() && !isPageLoaded) {
				isPageLoaded = true;
				init();
			}
		};
		addEvent(window, 'load', change);
		addEvent(doc, 'load', change);
		addEvent(doc, 'readystatechange', change);
		change();	// If the page is already loaded, the events above won't be fired
	}
})(window, true);
